File: assets/js/react/components/UnsplashSearch.js
import React, { useEffect, useState, useRef } from 'react'
import _ from 'lodash'

import Modal from 'react-bootstrap/Modal'
import ModalBody from 'react-bootstrap/ModalBody'
import Form from 'react-bootstrap/Form'
import ModalHeader from 'react-bootstrap/ModalHeader'
import ModalTitle from 'react-bootstrap/ModalTitle'
import Button from 'react-bootstrap/Button'
import Media from 'react-bootstrap/Media'
import Card from 'react-bootstrap/Card'
import CardColumns from 'react-bootstrap/CardColumns'
import ModalFooter from 'react-bootstrap/ModalFooter'
import axios from 'axios'

import { createMedia } from '../utilities/api'
import { useStateValue } from '../state'

import LoaderButton from './LoaderButton'

export const UnsplashLogo = ({size, className, white}) => {
    const fillColor = white ? '#fff' : '#000000'
    return <svg className={className} width={size} xmlns="http://www.w3.org/2000/svg" viewBox="0 0 400 400"><g><g><polygon style={{ fill: fillColor }} points="274.14 177.12 400 177.12 400 400 0 400 0 177.12 125.86 177.12 125.86 288.33 273.68 288.33 273.68 177.12 274.14 177.12"/><polygon style={{ fill: fillColor }} points="274.14 0 125.86 0 125.86 111.67 273.68 111.67 273.68 0 274.14 0"/></g></g></svg>
}

UnsplashLogo.defaultProps = { size: 18, white: false }

const UnsplashImageCard = ({ image, selected, handleSelect}) => {
    selected = selected ? 'bdo-unsplashSearch__card--selected' : ''

    return (
        <Card className={`bdo-unsplashSearch__card ${selected}`} onClick={() => handleSelect(image)}>
            <Card.Img variant="top" src={image.urls.small} />
            <Card.ImgOverlay className="d-flex align-items-end" style={{ background: 'linear-gradient(10deg, rgba(0,0,0,0.4), transparent)' }}>
                <Card.Text>
                    <Media className="text-light align-items-center">
                        <img src={image.user.profile_image.small} alt={image.user.name} className="rounded-circle mr-3" />
                        <Media.Body >{image.user.name}</Media.Body>
                    </Media>
                </Card.Text>
            </Card.ImgOverlay>
        </Card>
        // <div className="bdo-UnsplashSearch__item">
        //     <img src={image.urls.small} alt={image.description} className="bdo-UnsplashSearch__thumbnail" />
        //     <div className="bdo-UnsplashSearch__overlay">
        //         <div>
        //             <img src={image.user.profile_image.small} alt={image.user.name} />
        //             <span>{image.user.name}</span>
        //         </div>
        //     </div>
        // </div>
    )
}

const UnsplashSearch = ({ selectedImage, setSelectedImage }) => {
    const [inputValue, setInputValue] = useState('')
    const [query, setQuery] = useState('')
    const [page, setPage] = useState(1)
    const [results, setResults] = useState([])

    useEffect(() => {
        if (!query)
            return

        fetch(`https://api.unsplash.com/search/photos?page=${page}&query=${query}&client_id=oEPr7PwuQW8HmQB7q6JPdzC5y2pxQtnErE5zuZO1lTY`)
            .then(res => res.json())
            .then(json => setResults(json.results))
    }, [query])

    useEffect(() => {
        if (!query)
            return

        fetch(`https://api.unsplash.com/search/photos?page=${page}&query=${query}&client_id=oEPr7PwuQW8HmQB7q6JPdzC5y2pxQtnErE5zuZO1lTY`)
            .then(res => res.json())
            .then(json => setResults([...results, ...json.results]))
    }, [page])

    const delayedQuery = useRef(_.debounce(q => setQuery(q), 500)).current

    const handleSearch = e => {
        const value = e.target.value
        setInputValue(value)
        delayedQuery(value)
    }

    return (
        <div>
            <Form.Control type="text" placeholder="Search free high-resolution photos" size="lg" className="mb-3" value={inputValue} onChange={handleSearch} />
            <div style={{maxHeight: '73vh', overflow: 'scroll'}} className="p-1">
                <CardColumns style={{ maxHeight: page * 800 }}>
                    {results.map(image => {
                        let selected = !selectedImage ? null : selectedImage.id === image.id
                        return (
                            <UnsplashImageCard
                                image={image}
                                handleSelect={image => setSelectedImage(image)}
                                selected={selected}
                            />
                        )
                    })}
                </CardColumns>
                {!results.length ? null : <div className="text-center">
                    <Button
                        variant="outline-secondary"
                        onClick={() => setPage(page + 1)}
                    >
                        Load More
                    </Button>
                </div>}
            </div>
        </div>
    )
}

export const UnsplashButton = ({children}) => {
    const [showModal, setShowModal] = useState(false)

    const childElement = children ? React.cloneElement(
        children,
        {
            onClick: () => setShowModal(true)
        }
    ) : null

    return (
        <>
            <UnsplashSearchModal
                show={showModal}
                handleShow={setShowModal}
                onSelectImage={null}
            />
            {childElement}
        </>
    )
}

const UnsplashSearchModal = ({ show, handleShow }) => {
    const [{activeBrand, assets}, dispatch] = useStateValue()
    const [selectedImage, setSelectedImage] = useState(null)

    const handleUpload = image => {
        let formData = new FormData()
        formData.append('upload_preset', `bdo_image_frontend_upload`)
        formData.append('file', image.urls.full)

        axios.request({
            method: 'POST',
            url: `https://api.cloudinary.com/v1_1/bordo/upload`,
            data: formData
        })
            .then(res => {
                createMedia(activeBrand.id, image.description || image.id, res.data)
                    .then(data => {
                        dispatch({
                            type: 'setAssets',
                            assets: [data, ...assets]
                        })

                        handleShow()
                    })
            })
    }

    return (
        <Modal size="lg"
            show={show}
            onHide={handleShow}
            handleShow={handleShow}
            handleCl
            centered
        >
            <ModalHeader closeButton>
                <ModalTitle><UnsplashLogo size={32} /> Unsplash</ModalTitle>
            </ModalHeader>
            <ModalBody>
                <UnsplashSearch
                    selectedImage={selectedImage}
                    setSelectedImage={setSelectedImage}
                />
            </ModalBody>
            {!selectedImage ? null : <ModalFooter><LoaderButton variant="success" onClick={() => handleUpload(selectedImage)}>Select Image</LoaderButton></ModalFooter>}
        </Modal>
    )
}

export default UnsplashSearchModal

File: lib/bordo_web/live/unsplash_search.ex
defmodule BordoWeb.Media.UnsplashSearch do
  use BordoWeb, :live_component

  @client_id "oEPr7PwuQW8HmQB7q6JPdzC5y2pxQtnErE5zuZO1lTY"

  def render(assigns) do
    ~L"""
      <div id="unsplash-search">
        <%= if @loading do %>
          <div class="bg-gray-50 overflow-hidden rounded-lg">
            <div class="px-4 py-5 sm:p-6">
              <span class="spinner"></span><span class="ml-3">Uploading...<span>
            </div>
          </div>
        <% else %>
          <form phx-submit="unsplash-search" phx-target="#unsplash-search" class="mb-4">
            <input name="search" value="<%= @unsplash_search %>" class="bg-white focus:outline-none focus:shadow-outline border border-gray-300 rounded-lg py-2 px-4 block w-full appearance-none leading-normal" autocomplete="off" autofocus placeholder="Search Unsplash" %>
            <p class="mt-2 text-xs text-gray-500" id="email-description">Press Enter to search.</p>
          </form>
          <%= if Enum.any?(@unsplash_results) do %>
            <div class="card-columns mb-4">
              <%= for result <- @unsplash_results do %>
                <%= image_card(result) %>
              <% end %>
            </div>
            <div class="grid grid-cols-2">
              <div class="cursor-pointer" phx-click="nav" phx-value-page="<%= Enum.max([1, (@page - 1)]) %>" phx-target="#unsplash-search">&larr; Previous Page</div>
              <div class="text-right cursor-pointer" phx-click="nav" phx-value-page="<%= @page + 1 %>" phx-target="#unsplash-search">Next Page &rarr;</div>
            </div>
          <% else %>
            <div class="bg-gray-50 overflow-hidden rounded-lg">
              <div class="px-4 py-5 sm:p-6">
                No results, try a different search
              </div>
            </div>
          <% end %>
        <% end %>
      </div>
    """
  end

  def handle_event("unsplash-search", %{"search" => search}, socket) do
    {:noreply,
     socket
     |> assign(:unsplash_results, fetch_results(search, socket.assigns.page))
     |> assign(:unsplash_search, search)}
  end

  def handle_event("image-selected", %{"url" => url}, socket) do
    send(self(), {:upload, url})
    {:noreply, socket |> assign(:loading, true)}
  end

  def handle_event("nav", %{"page" => page}, socket) do
    {:noreply,
     socket
     |> assign(
       :unsplash_results,
       fetch_results(socket.assigns.unsplash_search, page)
     )
     |> assign(:page, String.to_integer(page))}
  end

  def image_card(image_data) do
    ~e"""
      <div class="inline-block rounded-sm relative cursor-pointer overflow-hidden" phx-click="image-selected" phx-value-url="<%= image_data["urls"]["full"] %>" phx-target="#unsplash-search">
        <img src="<%= image_data["urls"]["small"] %>">
        <div class="flex items-end absolute top-0 left-0 right-0 bottom-0 p-3" style="background: linear-gradient(10deg, rgba(0, 0, 0, 0.4), transparent);">
          <div class="text-gray-50 items-center flex">
            <img src="<%= image_data["user"]["profile_image"]["small"] %>" alt="<%= image_data["user"]["name"] %>" class="rounded-circle mr-3" />
            <div class="flex-1"><%= image_data["user"]["name"] %></div>
          </div>
        </div>
      </div>
    """
  end

  def fetch_results(search, page) do
    HTTPoison.get!(
      "https://api.unsplash.com/search/photos?page=#{page}&query=#{URI.encode(search)}&client_id=#{
        @client_id
      }"
    )
    |> Map.get(:body)
    |> Jason.decode!()
    |> Map.get("results")
  end
end
